### DISCUSSION TITLE
Is it possible to "Pick" a subset of properties into a model?

### DISCUSSION THREAD

**Discussion by elliots** [microsoft/typespec#3484] [Ideas]
The usecase is that we have models that are used by different apis, sometimes with only a subset of the properties. For example, emitting CustomerCreated events to a webhook that only include id, name, email from Customer.

I could create BaseCustomer and have multiple things extend it, but maybe there's another way?

e.g. 
```typespec
model Widget {
  @key id: string;
  weight: int32;
  color: "red" | "blue";
}

model WidgetNoColor extends Pick<Widget, "id" | "weight"> {
}

model WidgetNoWeight extends Widget {
  weight: never;
}

model WidgetCreatedEvent {
  widgetId: Widget["id"]
}
```

**Comment by bterlson** [comment]
We don't have Pick yet unfortunately. We have Omit, though it's done with a mutating decorator which is questionably good. That said, you could look at [how that works](https://github.com/microsoft/typespec/blob/main/packages/compiler/lib/std/types.tsp#L51) and implement something similar for Omit. Alternatively, you can pull out e.g. id and weight into a separate model and spread it in where needed.

**Reply by elliots** [reply]
I'll check out Omit (which might actually work for some uses) and see how a Pick might operate. But splitting out and using as needed is probably most sensible. Thankyou!

**Reply by btiernay** [reply]
Was there any consideration given to supporting both `Omit` and `Pick` as first class utility types?

**Reply by timotheeguerin** [reply]
do you mean without the decorator patch, as a more built-in type? If so I think it would be nice but for now those types do make it work well without a design on how those special types would work/new conditional syntax like TypeScript.

**Reply by btiernay** [reply]
> without a design on how those special types would work/new conditional syntax like TypeScript
Can you please clarify what you mean regarding "conditional syntax"?  

The intuition for my question is that `Omit`, `Pick` and others are very frequently used type operations and would probably benefit from first class support in ASTs, etc. I was curious if these were ever considered.

**Comment by elliots** [comment]
The `Pick` PR was accepted and released in 0.57 - https://github.com/microsoft/typespec/pull/3488
